[mlir][spirv] Support poison index when converting vector.insert/extract

andfau-amd opened the pull request:
This modifies the conversion patterns so that, in the case where the index is known statically to be poison, the insertion/extraction is replaced by an arbitrary junk constant value, and in the dynamic case, the index is sanitized at runtime. This avoids triggering a UB in both cases. The dynamic case is definitely a pessimisation of the generated code, but the use of dynamic indexes is expected to be very rare and already slow on real-world GPU compilers ingesting SPIR-V, so the impact should be negligible.

Resolves #124162.

andfau-amd commented on mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:
```diff
@@ -154,14 +174,26 @@ struct VectorExtractOpConvert final
-          rewriter.getI32ArrayAttr(id.value()));
-    else
+            getConstantIntValue(extractOp.getMixedPosition()[0])) {
+      // TODO: It would be better to apply the ub.poison folding for this case
+      //       unconditionally, and have a specific SPIR-V lowering for it,
+      //       rather than having to handle it here.
```
Any thoughts on this? (Same comment exists for the other op also).

If nobody objects, I'll make an issue for this, so I can link it in the comments.

andfau-amd replied:
I was assuming this thing didn't try to fold first because otherwise the test would be failing, but looking at that code snippet, it seems I hadn't considered the possibility it actually folds _second_ (if legalization fails). I should see what happens if I make it reject the poison index instead.

andfau-amd replied:
It seems like if I make this conversion pattern reject the poison index, the operation just never gets legalized (looking at the unit test output). I guess this means the folder isn't aware of or for some other reason doesn't apply the poison folding pattern; I'm not yet sure why.

andfau-amd replied:
Ahh, no: `legalizeWithFold` calls `tryFold`, `tryFold` calls `ExtractOp::fold`, `ExtractOp::fold` calls `foldPoisonIndexInsertExtractOp`, and a poison attribute is generated, but then `tryFold` calls `VectorDialect::materializeConstant` and somewhere in there, the poison is rejected. Apparently, `VectorDialect::materializeConstant` calls `arith::ConstantOp::materialize`; maybe I just need to add a path to it that calls the UB dialect equivalent?

kuhar replied:
This should be fixed with https://github.com/llvm/llvm-project/pull/125596 . I think we can wait for this to land first.

andfau-amd replied:
Oh, I didn't know about that PR! I ended up implementing basically the same thing, but it turns out that's not enough by itself to fix things for `-convert-vector-to-spirv`. Presumably the SPIR-V dialect is rejecting the `ub.poison` after it's created or something along these lines, I need to keep digging.

kuhar replied:
We don't have ub.poison conversion implemented yet: https://github.com/llvm/llvm-project/pull/125613#pullrequestreview-2591515349. But in the vector conversion tests it can be left unconverted for the time being.

kuhar commented on mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:
```diff
@@ -137,6 +137,26 @@ struct VectorBroadcastConvert final
+      rewriter.getIntegerAttr(dynamicIndex.getType(), kPoisonIndex));
+  Value cmpResult =
+      rewriter.create<spirv::IEqualOp>(loc, dynamicIndex, poisonIndex);
+  Value sanitizedIndex = rewriter.create<spirv::SelectOp>(
+      loc, cmpResult,
+      spirv::ConstantOp::getZero(dynamicIndex.getType(), loc, rewriter),
```
Selects are fairly expensive on many GPU targets. Instead, Can we clamp power-of-two vectors using an `and` with a mask? I'd expect that any performance-oriented code uses pow2 vectors.

andfau-amd commented on mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:
```diff
@@ -137,6 +137,26 @@ struct VectorBroadcastConvert final
+// choosing to replace the poison index with index 0 (always in-bounds).
+static Value sanitizeDynamicIndex(ConversionPatternRewriter &rewriter,
+                                  Location loc, Value dynamicIndex,
+                                  int64_t kPoisonIndex) {
```
There's a separate `kPoisonIndex` constants for `ExtractOp` and `InsertOp`, and it feels unclean to assume they have the same value, even if that happens to be the case in practice.

kuhar replied:
I think that both are defined here: https://github.com/llvm/llvm-project/blob/84fbed86ffcb97c24f9294a204c60da5444b8646/mlir/include/mlir/Dialect/Vector/IR/Vector.td#L36 but, like you noted, this is not something we can universally query. Let's leave it as-is. Thanks!

andfau-amd committed:
[mlir][spirv] Support poison index when converting vector.insert/extract

This modifies the conversion patterns so that, in the case where the
index is known statically to be poison, the insertion/extraction is
replaced by an arbitrary junk constant value, and in the dynamic case,
the index is sanitized at runtime. This avoids triggering a UB in both
cases. The dynamic case is definitely a pessimisation of the generated
code, but the use of dynamic indexes is expected to be very rare and
already slow on real-world GPU compilers ingesting SPIR-V, so the impact
should be negligible.

Resolves #124162.

andfau-amd commented on mlir/lib/Conversion/VectorToSPIRV/VectorToSPIRV.cpp:
```diff
@@ -154,14 +181,26 @@ struct VectorExtractOpConvert final
-    else
+            getConstantIntValue(extractOp.getMixedPosition()[0])) {
+      // TODO: ExtractOp::fold() already can fold a static poison index to
+      //       ub.poison; remove this once ub.poison can be converted to SPIR-V.
```
(I intend to tackle ub.poison to SPIR-V conversion myself soon, and I'll clean this up in that commit.)

kuhar reviewed (approved):
LGTM
